### PR TITLE
Enhance model builder tool for building and running models

### PR DESCRIPTION
**Pull request by kunal-vaishnavi**
### Description

This PR enhances the model builder tool for building and running models by adding the following.

- Adds a README file to show how to use the tool
- Creates a GenAI JSON config file that users can customize
- Saves the tokenization files from Hugging Face
- Adds additional options for customizing how the model is built
- Allows mini models to be created for testing purposes
- Rewrites `create_model` for integration with Olive

### Motivation and Context

- The README file showcases the tool's full usage.
- The JSON file has two groups, `model` and `search`. The `model` section contains information about the model and the parameters in the `search` section can be modified as desired.
- The tokenization files are needed for the tokenizers in the ORT GenAI library.
- The full list of additional options can be viewed in the usage section of the README.
- By reducing the number of hidden layers, mini models can be created with the model builder tool for testing purposes (e.g. unit testing, CIs).
- [This PR](https://github.com/microsoft/Olive/pull/940) is integrating the model builder tool into Olive and wants to import the `create_model` method directly instead of running the script.

### CODE REVIEW

File: src/python/setup.py.in [hunk 1]
```diff
@@ -13,7 +13,7 @@ class BinaryDistribution(Distribution):
 setup(
     name='@TARGET_NAME@',
     version='@VERSION_INFO@',
-    packages=find_packages(),
+    packages=['onnxruntime_genai', 'onnxruntime_genai.models'],
```
Comment (natke) [inline src/python/setup.py.in:16]: Should this just be onnxruntime_genai?

Comment (kunal-vaishnavi) [reply src/python/setup.py.in:16]: If `onnxruntime_genai.models` is not listed, then the `models` folder does not appear in the wheel.

File: src/python/py/models/README.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+2. Modify `num_hidden_layers` in `config.json` to your desired target (e.g. 4 layers).
+3. Run the below command for the model builder tool.
+
+```
+# From wheel:
+python -m onnxruntime_genai.models.builder -m model_name -o /path/to/output/folder -p precision -e execution_provider -c cache_dir_where_hf_files_are_saved
+
+# From source:
+python builder.py -m model_name -o /path/to/output/folder -p precision -e execution_provider -c cache_dir_where_hf_files_are_saved
+```
```
Comment (yufenglee) [inline src/python/py/models/README.md:103]: Coud you please add a section to explain how user can modify the script and support their customized model?

File: src/python/py/models/builder.py [hunk 3]
```diff
@@ -108,8 +116,58 @@ def __init__(self, config, io_dtype, onnx_dtype, ep, cache_dir, extra_options):
...
+                "diversity_penalty": config.diversity_penalty if hasattr(config, "diversity_penalty") else 0.0,
+                "length_penalty": config.length_penalty if hasattr(config, "length_penalty") else 1.0,
+                "max_length": config.max_length if hasattr(config, "max_length") and config.max_length != hf_default_search_values["max_length"] else self.context_length,
+                "min_length": 0,
+                "no_repeat_ngram_size": config.no_repeat_ngram_size if hasattr(config, "no_repeat_ngram_size") else 0,
+                "num_beams": config.num_beams if hasattr(config, "num_beams") else 1,
+                "num_return_sequences": config.num_return_sequences if hasattr(config, "num_return_sequences") else 1,
+                "repetition_penalty": config.repetition_penalty if hasattr(config, "repetition_penalty") else 1.0,
+                "temperature": config.temperature if hasattr(config, "temperature") else 1.0,
+                "top_k": config.top_k if hasattr(config, "top_k") and config.top_k != hf_default_search_values["top_k"] else 0,
```
Comment (yufenglee) [inline src/python/py/models/builder.py:154]: why do we need to check "config.top_k != hf_default_search_values["top_k"]"? 
With the check, top_k will be set to 0 if the top_k in config is same as hf_default_search_values["top_k"]

Comment (natke) [reply src/python/py/models/builder.py:154]: If the top k is specified by the HuggingFace config, we want to copy that value into our config.
If the top k is not specified by the HuggingFace config, we want to set to a default value of 50

Comment (natke) [reply src/python/py/models/builder.py:154]: Re: max length:

If the maximum length is specified by the HuggingFace config, we want to copy that value into our config
If the maximum length is not specified by the HuggingFace config, we want to set to a default value of the model max length

**Review by RyanUnderhill — APPROVED** [review]

File: src/python/py/models/README.md (hunk 2 above)
Comment (kunal-vaishnavi) [reply src/python/py/models/README.md:103]: I'll add it in another PR

**Pull request merged**